signer/xpi/x509.go

```diff
@@ -61,33 +62,63 @@ func (s *XPISigner) monitorRsaCacheSize() {
 	}
 }
 
-// retrieve a key from the cache or generate one if it takes too long
-// or if the size is wrong
+type rsaKey struct {
+	lock  sync.Mutex
+	age   time.Time
+	usage int
+	key   *rsa.PrivateKey
+}
+
+var currentRsaKey rsaKey
+
+// getRsaKey applies some intelligence to key management. It will return the
+// current key if its lifetime or usage count aren't exceeded, or will get a
+// new key from the cache. If the cache is unresponsive, it will generate a new
+// RSA key by itself and return it. All while publishing stats so we can
+// monitor the state of the cache.
 func (s *XPISigner) getRsaKey(size int) (*rsa.PrivateKey, error) {
 	var (
-		err error
-		key *rsa.PrivateKey
+		err   error
 		start time.Time
 	)
+
+	// we're messing with pointers and counters shared across goroutines, so
+	// only allow one execution of this function at a time
+	currentRsaKey.lock.Lock()
+	defer currentRsaKey.lock.Unlock()
+
+	// see if we can reuse the current key
+	if currentRsaKey.key != nil &&
+		currentRsaKey.age.Add(s.rsaKeyMaxAge).After(time.Now()) && // if current key hasn't reached max lifetime
+		currentRsaKey.usage < s.rsaKeyMaxUsage { // if current key hasn't reached max usage
+		currentRsaKey.usage++
+		return currentRsaKey.key, nil
```

Comment:
nit: we should either hoist `start` to above the lock and send to the histogram from this early exit or remove sending it later since it'll be incomplete